### PR TITLE
Extend SlurmSpawner batch_script to include `.condarc` customization

### PR DESCRIPTION
**Pull request by viniciusdc**
## Reference Issues or PRs

#147 

## What does this implement/fix?

Extends ` QhubHPCSpawnerBase` class to accommodate a callable object instead of only strings to help support custom ` batch_script`  definitions that require context information from the Jupyterhub `Spawner` -- This is a step towards future enhancements on the dynamic config manipulation. Additionally, the formatting of the code was improved for better readability.

_Put a `x` in the boxes that apply_

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds a feature)
- [ ] Breaking change (fix or feature that would cause existing features not to work as expected)
- [ ] Documentation Update
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build related changes
- [ ] Other (please describe):

## Testing

- [ ] Did you test the pull request locally?
- [ ] Did you add new tests?

#### Text styling

- [ ] The content is written with [plain language](https://www.plainlanguage.gov/guidelines/) (where relevant).
- [ ] If there are headers, they use the proper header tags (with only one level-one header: `H1` or `#` in markdown).
- [ ] All links describe where they link to (for example, check the [Nebari website](https://nebari.dev/)).
- [ ] This content adheres to the Nebari style guides.

#### Non-text content

- [ ] All content is represented as text (for example, images need alt text, and videos need captions or descriptive transcripts).
- [ ] If there are emojis, there are not more than three in a row.
- [ ] Don't use [flashing GIFs or videos](https://www.w3.org/TR/UNDERSTANDING-WCAG20/seizure-does-not-violate.html).
- [ ] If the content were to be read as plain text, it still makes sense, and no information is missing.

### CODE REVIEW

File: roles/jupyterhub/templates/jupyterhub_config.py [hunk 1]
```diff
@@ -119,74 +124,95 @@ async def authenticate(self, handler, data):
...
-    ## for dashboards, since already selected a conda env
-    #if spawner.orm_spawner.user_options and 'presentation_type' in spawner.orm_spawner.user_options:
-    #  self.log.info("In options_form")
-    #  if spawner.user_options['presentation_type']:
-    #    return self.main_options_form # Omit the conda env dropdown since that is chosen per dashboard
 
-    # Display full form including conda env dropdown
-    return ''.join([self.main_options_form, self.conda_options_form])
+
+{% if jupyterhub_qhub_options_form %}
```
Comment (viniciusdc) [inline roles/jupyterhub/templates/jupyterhub_config.py:150]: This part of the `jupyterhub_qhub_form` only changed indentation as it was incorrect, no other changes were made here

File: roles/jupyterhub/templates/jupyterhub_config.py [hunk 2]
```diff
@@ -119,74 +124,95 @@ async def authenticate(self, handler, data):
...
+        return ''.join([self.main_options_form, self.conda_options_form])
 {% endif %}
 
 # Assign Qhub Spawner
 class QHubHPCSpawner(QHubHPCSpawnerBase):
     pass
 
+c.JupyterHub.allow_named_servers = True
+c.JupyterHub.default_url = '/hub/home'
+
```
Comment (viniciusdc) [inline roles/jupyterhub/templates/jupyterhub_config.py:215]: I am not too sure where this came from, can be removed if necessary, but did not affect the instance creation

**Comment by viniciusdc** [comment]
Tested locally with ubuntu2004 test box

image

**Comment by aktech** [comment]
I am seeing this, when running on the cluster:

```

RUNNING HANDLER [jupyterhub : Restart services jupyterhub] ***********************************************************************************************************************************
failed: [jhub-login] (item=jupyterhub) => {"ansible_loop_var": "item", "changed": false, "item": "jupyterhub", "msg": "Unsupported parameters for (ansible.legacy.systemd) module: cmd. Supported parameters include: daemon_reexec, daemon_reload, enabled, force, masked, name, no_block, scope, state (daemon-reexec, daemon-reload, service, unit)."}

PLAY RECAP ***********************************************************************************************************************************************************************************
jhub-compute-01            : ok=64   changed=0    unreachable=0    failed=0    skipped=36   rescued=0    ignored=0
jhub-compute-02            : ok=63   changed=0    unreachable=0    failed=0    skipped=35   rescued=0    ignored=0
jhub-login                 : ok=195  changed=6    unreachable=0    failed=1    skipped=53   rescued=0    ignored=0
```

**Comment by viniciusdc** [comment]
This was due to the changes made by the linter, it add that `cmd` entry under all handlers. Check #150

**Comment by aktech** [comment]
Ah, I see make sense. I think we can merge that first.

**Comment by viniciusdc** [comment]
Hi @aktech, sorry I really don't see any issues with this as of now, I will do another deployment later today to double check but I would be curious to see if you can inspect if the jupyterhub.py config file was indeed updated after the ansible task has run... could ansible have skipped the task?

File: roles/jupyterhub/templates/jupyterhub_config.py (hunk 2 above)
Comment (aktech) [reply roles/jupyterhub/templates/jupyterhub_config.py:215]: We need to remove anything that isn't required to fix the underlying issue.

**Comment by viniciusdc** [comment]
> That's alright, I can test that. Meanwhile, can you undo formatting / refactor changes in this PR to keep the diff minimal, and we can do cleanup, refactor in a follow up PR?

Sure thing, just keep in mind when you start testing that some of those changes are required to do the provision correctly

**Comment by aktech** [comment]
> some of those changes are required to do the provision correctly

Which changes? linting? changes that should be a separate PR and merged first separately.

**Comment by viniciusdc** [comment]
> > some of those changes are required to do the provision correctly
> 
> Which changes? linting? changes that should be a separate PR and merged first separately.

nvm, I got confused with the other PRs hshshsh -- there are not much to remove from this as mostly of the changes here were made so that we could use the spawner class attributes, which required python function to load

**Review by aktech — APPROVED** [review]
This works! thanks for working on this. 🎉

**Pull request merged**